tremor-script/src/ast/optimizer.rs

```diff
@@ -0,0 +1,80 @@
+use crate::ast::visitors::{ArrayAdditionOptimizer, ConstFolder};
+use crate::ast::{
+    CreationalWith, DefinitionalArgs, Expr, FnDefn, Helper, ImutExpr, Query, Script,
+    ScriptDefinition, SelectStmt, Stmt, WindowDefinition,
+};
+use crate::errors::Result;
+
+/// Runs optimisers on a given AST element
+pub struct Optimizer<'run, 'script, 'registry>
+where
+    'script: 'run,
+{
+    helper: &'run Helper<'script, 'registry>,
+}
+
+// Implemented as a macro, so we don't have to move the optimizers into an array/vec where they'd
+// have to be Boxed
+macro_rules! walk_thing(
+    ($name:ident, $ty:ident, { $additional_uses:stmt }) => {
+        walk_thing!($name, $ty, $name, { $additional_uses });
+    };
+    ($name:ident, $ty:ident, $walker_name:ident, { $additional_uses:stmt }) => {
+        /// Walk $name
+        /// # Errors
+        /// Fails if the optimizer function fails
+        pub fn $name(self, walkable: &mut $ty<'script>) -> Result<()> {
+            $additional_uses
+
+            ConstFolder::new(self.helper).$walker_name(walkable)?;
+            ArrayAdditionOptimizer{}.$walker_name(walkable)?;
+
+            Ok(())
+        }
+    }
+);
+
+impl<'run, 'script, 'registry> Optimizer<'run, 'script, 'registry>
```

Comment:
I think a nicer approach would be something like:
```rust
 CombinedVisitor<First, Second>
  {
  first: First,
  second: Second
}

impl<First, Second> ImutExprWalker for CombinedVisitor
  where
  First: ImutExprVisitor,
  Second: ImutExprVisitor,
{}
impl<First, Second> ImutExprVisitor for CombinedVisitor
  where
  First: ImutExprVisitor,
  Second: ImutExprVisitor,
{
/// ... all the impls ... his is where a macro could go :P
    fn leave_expr(&mut self, e: &mut ImutExpr<'script>) -> Result<()> {
      first.leave_expr(e)?;
      second.leave_expr(e)?;
    }
/// ... all the impls ...
}
/// same for Deploy, Expr, Query


pub struct Optimizer<'run, 'script, 'registry>
where
    'script: 'run,
{
    visitor: CombinedVisitor<ConstFolder<'run, 'script, 'registry>, ArrayAdditionOptimizer>
}

 impl<'run, 'script, 'registry> Optimizer<'run, 'script, 'registry>{
    pub fn new(helper: &'run Helper<'script, 'registry>) -> Result<Self> {
        Ok(Self { visitor: CombinedVisitor::new(ConstFolder::new(self.helper).$walker_name(walkable)?, ArrayAdditionOptimizer{})}
    })
}

impl<'run, 'script, 'registry> AsRef<CombinedVisitor<ConstFolder<'run, 'script, 'registry>, ArrayAdditionOptimizer>
}> for Optimizer<'run, 'script, 'registry>{
fn as_ref(&self) -> &CombinedVisitor<ConstFolder<'run, 'script, 'registry>, ArrayAdditionOptimizer>
&self.visitor
}
}
}
```

Reply:
How would that approach work when more optimizers get added? Adding more generic arguments seems very tedious.

Reply:
Absolutely, two are enough:) they can be chained ;). As the combiner also is a visitor it can combine itself. 

One advantage is that it only needs to walk once no matter how many visitors are present